jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -571,61 +574,67 @@ static Chunk next(Chunk chunk)
 
         /**
          * <p>Returns a new {@code Chunk} whose {@code ByteBuffer} is a slice of the
-         * {@code ByteBuffer} of the source {@code Chunk} unless the source
-         * {@link #hasRemaining() has no remaining byte} in which case:</p>
+         * whole {@code ByteBuffer} of the source {@code Chunk} unless the source
+         * {@link #hasRemaining() has no remaining bytes} in which case:</p>
          * <ul>
-         * <li>{@code this} is returned if it is an instance of {@link Error}</li>
-         * <li>{@link #EOF} is returned if {@link #isLast()} is {@code true}</li>
+         * <li>{@code this} is returned if {@link #isLast()} is {@code true}</li>
          * <li>{@link #EMPTY} is returned if {@link #isLast()} is {@code false}</li>
          * </ul>
-         * <p>If the source has remaining bytes, the returned {@code Chunk} retains
-         * the source {@code Chunk} and it is linked to it via
+         * <p>If the source {@code Chunk} has remaining bytes, the returned
+         * {@code Chunk} it is linked to the source {@code Chunk} via
          * {@link #from(ByteBuffer, boolean, Retainable)}.</p>
          *
-         * @return a new {@code Chunk} retained from the source {@code Chunk} with a slice
-         * of the source {@code Chunk}'s {@code ByteBuffer}
+         * @return a new {@code Chunk} with a slice of the source {@code Chunk}'s
+         * {@code ByteBuffer}
          */
         default Chunk slice()
         {
-            if (isTerminal())
-                return this;
-            if (!hasRemaining())
-                return EMPTY;
-            retain();
-            return from(getByteBuffer().slice(), isLast(), this);
+            if (hasRemaining())
+                return from(getByteBuffer().slice(), isLast(), this);
+            else
+                return isLast() ? this : EMPTY;
```

Comment:
why not:
```suggestion
                return isLast() ? EOF : EMPTY;
```

Reply:
Because it can also be an instance of `Error`, so returning `this` is correct.

Reply:
But if this is just an empty normal buffer, then the pointers of the "slice" can be adjusted to see content around it. It will also change the original chunk.

Perhaps Error needs to implement slice

Reply:
I don't understand? 
We don't return empty Chunks that are not `EMPTY`, which has `capacity==0` so you cannot adjust position/limit.
In all other cases, we slice the `ByteBuffer` from the network buffer, so you can never see around what you're supposed to see.

Reply:
Consider if we have a chunk that we are parsing other chunks out of and slicing them to pass other processing.   That other processing should not be able to adjust the pointers to look at content that was not destined for it.   If the slice contains data, then they cannot do so as the code creates a new chunk from a slice.    But if the original chunk is empty and we slice off it, then we just get the original chunk.   So problem 1 is if we release that, then we are releasing the original. Second problem is that "slice" can have it's pointers adjusted to look at data it should not be able to see, which also modifies the original chunk.

A slice should be a slice and not `this`

Reply:
> But if the original chunk is empty and we slice off it, then we just get the original chunk.

No, you get `EMPTY`.

`this` is only returned if it is empty _and_ last, to cover the cases of `Error` and `Trailers` (and it's just that we have to return something, as no code will ever slice an empty last chunk).

A slice of an empty and last chunk is itself, and you cannot move position/limit anyway even in the original chunk, so there is no point in returning anything different from `this`.

Reply:
So if it was a slice of an empty last normal chunk then returning this is wrong. Return EOF here and extend Error to return this from slice.

Reply:
Also resolving this conversation to be replaced by new review conversation.